### PR TITLE
Add entity system for complex objects and animations with Aseprite export/import

### PR DESCRIPTION
**Pull request by justinvh**
This changeset adds an animation system to the game, enabling the representation of complex sprites, including sprites with multiple layers and many OAMs. The system allows you to export animations from Aseprite, which can then be imported into the game at compile time using a custom header syntax.

Animation System TODO
---------------------
:white_check_mark: Sprite, Animation, Frame, and Tile headers
:white_check_mark: Export of animations from Aseprite
:white_check_mark: Load sprites from Asesprite export into VRAM
:white_check_mark: Load a tag and associated OAM objects. Render to screen.
:white_check_mark: Load mixes of big and small tiles and let the sprite optimizer choose tiles
:white_check_mark: Iterate through each tag and render all frames

Results
-------
complex-sprite-animation-5

Screenshot 2023-07-16 213237

Example output from Aseprite Exporter
----------------------------
```
python .\asesprite2bin.py -i ..\resources\sprites\plane -b 3
INFO:build:Reading Aseprite sprite from D:\projects\top-gun-3\resources\sprites\plane
INFO:build:Reading Asesprite JSON data from D:\projects\top-gun-3\resources\sprites\plane\plane.json
INFO:build:Building sprite header from dictionary
INFO:build:Sprite has 23 frames
INFO:build:Calculating OAM tiles for each layer
INFO:build:     Layer Plane with 4 big tiles and 0 small tiles
INFO:build:     Layer Exhaust with 0 big tiles and 12 small tiles
INFO:build:     Layer Vapor with 6 big tiles and 0 small tiles
INFO:build:Loading sprite sheet plane.png
INFO:build:     Loaded 512 bytes from D:\projects\top-gun-3\resources\sprites\plane\plane.pal
INFO:build:     Loaded 6656 bytes from D:\projects\top-gun-3\resources\sprites\plane\plane.bin
INFO:build:Building tags
INFO:build:     Tag 1: Forward with 1 frames (dir=forward)
INFO:build:     Tag 2: Forward-Afterburner with 3 frames (dir=forward)
INFO:build:     Tag 3: Supersonic with 3 frames (dir=forward)
INFO:build:     Tag 4: Bank with 1 frames (dir=forward)
INFO:build:     Tag 5: Bank-Afterburner with 3 frames (dir=forward)
INFO:build:Done building sprite header
INFO:build:PAL data is at offset 0
INFO:build:Sprite sheet data is at offset 512
INFO:build:Tag metadata is at offset 7168
INFO:build:Tag data is at offset 7193
INFO:serialize:Wrote sprite header to D:\projects\top-gun-3\resources\sprites\plane\plane.sprite (8483 bytes)
INFO:build:Wrote animation mapper to D:\projects\top-gun-3\resources\sprites\plane\plane.i (467 bytes)
```

Example loading sprites in the engine
-------------------------------------
```
; Load a demo sprite
jsr SpriteManager_Request
lda #Sprite_Plane@Bank
ldx #Sprite_Plane@Data
jsr Sprite_Load 

; Set the tag of the sprite to the Forward animation
lda #Sprite_Plane@Tag@Forward_Afterburner
jsr Sprite_SetTag

; Set the frame of the sprite to 0
lda #0
jsr Sprite_SetFrame
```

### CODE REVIEW

File: engine/oam.asm [hunk 1]
```diff
@@ -564,8 +599,34 @@ OAMManager_Init:
...
+        pha
         jsr OAMObject_Init      ; Initialize the current OAM object (X register points to it)
-        sty oam_object.index, X ; Set the index of the object to its position in the array
+
+        A8
+        tya
+        sta oam_object.index, X ; Set the index of the object to its position in the array
+
+        ; Compute word offset into the OAM low table (16 bits per object)
+        tya
```
Comment (Mena2871) [inline engine/oam.asm]: do you need this extra tya?

File: engine/map.asm [hunk 2]
```diff
@@ -331,16 +332,35 @@ Map_LoadSprites
...
+    A16
+
+    ; Convert the palette index to a CGRAM address
+    ; Expects the Y register to be the palette index
+    jsr CGRAM_Index
+
+    ; Bounce the stack
+    ply
+    plx
+    phy
```
Comment (Mena2871) [inline engine/map.asm]: order of stack stuff in this subroutine

File: engine/oam.asm (hunk 1 above)
Comment (justinvh) [reply engine/oam.asm]: No, just keep the code organized in terms of having blocks of "prepare accumulator, do something with accumulator" but I will remove it

File: engine/map.asm (hunk 2 above)
Comment (justinvh) [reply engine/map.asm]: Good catch

**Pull request merged**